server/src/main/java/org/apache/iotdb/db/query/executor/fill/PreviousFill.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.iotdb.db.exception.StorageEngineException;
+import org.apache.iotdb.db.exception.query.QueryProcessException;
+import org.apache.iotdb.db.qp.physical.crud.RawDataQueryPlan;
+import org.apache.iotdb.db.query.context.QueryContext;
+import org.apache.iotdb.db.query.control.QueryResourceManager;
+import org.apache.iotdb.tsfile.file.metadata.TimeseriesMetadata;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.read.TimeValuePair;
+import org.apache.iotdb.tsfile.read.common.Path;
+import org.apache.iotdb.tsfile.read.filter.TimeFilter;
+import org.apache.iotdb.tsfile.read.filter.basic.Filter;
+import org.apache.iotdb.tsfile.read.filter.factory.FilterFactory;
+
+import java.io.IOException;
+import java.util.*;
+
+public class PreviousFill extends IFill {
+
+  private Path seriesPath;
+  private QueryContext context;
+  private long beforeRange;
+  private Set<String> allSensors;
+  private Filter timeFilter;
+
+  private boolean untilLast;
+
+  public PreviousFill(TSDataType dataType, long queryTime, long beforeRange) {
+    this(dataType, queryTime, beforeRange, false);
+  }
+
+  public PreviousFill(long beforeRange) {
+    this(beforeRange, false);
+  }
+
+
+  public PreviousFill(long beforeRange, boolean untilLast) {
+    this.beforeRange = beforeRange;
+    this.untilLast = untilLast;
+  }
+
+
+  public PreviousFill(TSDataType dataType, long queryTime, long beforeRange, boolean untilLast) {
+    super(dataType, queryTime);
+    this.beforeRange = beforeRange;
+    this.untilLast = untilLast;
+  }
+
+  @Override
+  public IFill copy() {
+    return new PreviousFill(dataType,  queryTime, beforeRange, untilLast);
+  }
+
+  @Override
+  void constructFilter() {
+    Filter lowerBound = beforeRange == -1 ? TimeFilter.gtEq(Long.MIN_VALUE)
+        : TimeFilter.gtEq(queryTime - beforeRange);
+    // time in [queryTime - beforeRange, queryTime]
+    timeFilter = FilterFactory.and(lowerBound, TimeFilter.ltEq(queryTime));
+  }
+
+  public long getBeforeRange() {
+    return beforeRange;
+  }
+
+  @Override
+  public void configureFill(
+      Path path, TSDataType dataType, long queryTime, Set<String> sensors, QueryContext context) {
+    this.seriesPath = path;
+    this.dataType = dataType;
+    this.context = context;
+    this.queryTime = queryTime;
+    this.allSensors = sensors;
+    constructFilter();
+  }
+
+  @Override
+  public TimeValuePair getFillResult()
+      throws IOException, QueryProcessException, StorageEngineException {
+    QueryDataSource dataSource =
+        QueryResourceManager.getInstance().getQueryDataSource(seriesPath, context, timeFilter);
+    // update filter by TTL
+    timeFilter = dataSource.updateFilterUsingTTL(timeFilter);
+    LastPointReader lastReader = new LastPointReader(
+        seriesPath, dataType, allSensors, context, dataSource, queryTime, timeFilter);
+
+    return lastReader.getTimeseriesLastPointXXX();
```

Comment:
XXX?

Reply:
Oh, that's a temporary name forgot to correct.. Fixed